### ISSUE TITLE
Have INotifyPropertyChanging implement a cancellation mechanism

### ISSUE THREAD

**Issue by weitzhandler** [dotnet/corefx#19627]
Hi,

The `INotifyPropertyChanging` will be way more productive if it would include a cancellation technique.

My suggestion would be have the `PropertyChangingEventArgs` class derive from `CancelEventArgs`. That's not a breaking change, and will give hell of a lot more productivity.

Proposed change ([Line 13](https://github.com/dotnet/corefx/blob/master/src/System.ObjectModel/src/System/ComponentModel/PropertyChangingEventArgs.cs#L13))

      namespace System.ComponentModel
      {
    -     public class PropertyChangingEventArgs : EventArgs
    +     public class PropertyChangingEventArgs : CancelEventArgs

[Commit](https://github.com/weitzhandler/corefx/commit/eea57d8206b36a70cca9c288ec3eacae55b2af43)

**Comment by adampursley** [comment]
Consider having OriginalValue and NewValue properties on the PropertyChangingEventArgs class, so that the handler for the event has the necessary context to be able to decide whether the event should be cancelled.

Following that same line of thought, consider also having PreviousValue and CurrentValue properties on the PropertyChangedEventArgs class.

**Comment by weitzhandler** [comment]
@karelz is this on the roadmap? I'm willing to be assigned and implement.

**Comment by karelz** [comment]
@weitzhandler it's not on roadmap, but that should not prevent you from grabbing any issue. We take any (high quality) contributions even if it is not something we plan to do ourselves anytime soon.

This one needs formal API proposal as next step -- see example in [API review process](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/api-review-process.md). For example the value / impact of the API change is not described here well (at least not to the point someone with zero area expertise can understand it from the description).

**Comment by weitzhandler** [comment]
OK @karelz I'll update the OP to include those. Thanks for your kind response.

**Comment by weitzhandler** [comment]
OK @karelz I'll update the OP to include those. Thanks for your kind response.

---
@adampursley 
Obviously this is a good idea and the clear benefit of it is that we don't have to retrieve the values using reflection, but since we can't force devs to provide those values, they might be null of being unset misleading the consumer to think it was set to null. We could introduce another readonly property set by its constructor stating whether the value is unset, but that adds a layer of complexity and makes it somewhat cumbersome.

**Comment by adampursley** [comment]
@weitzhandler 
Yeah, agreed that unfortunately we can't assume that values will be provided for those properties by the developer raising the event. I appreciate you giving it some thought.

In fact... I would also suggest that you give some consideration to whether you want to actually update the existing PropertyChangingEventArgs class or to perhaps instead define a new alternative class.

Consider this scenario: an existing framework utilizes PropertyChangingEventArgs in it's current form, which inherits from EventArgs; therefore we know that this framework can't be implementing any cancel behavior for that event. If the PropertyChangingEventArgs class is then updated to inherit from CancelEventArgs, then a consumer of that framework would see the Cancel property on the event args object and could try to make use of it, but assuming that no one has gone back to revise the framework to introduce any cancel behavior, then the framework will appear to the consumer to be not functioning correctly.

**Comment by weitzhandler** [comment]
@karelz
I've updated the [OP](https://github.com/dotnet/corefx/issues/19627#issue-227846040).

---
@adampursley 
Obviously this is a good idea and the clear benefit of it is that we don't have to retrieve the values using reflection, but since we can't force devs to provide those values, they might be null of being unset misleading the consumer to think it was set to null.
We could however introduce another readonly property set by its constructor stating whether the value is unset, that will adds a layer of complexity tho and will make it somewhat cumbersome.

**Comment by weitzhandler** [comment]
@karelz  @adampursley 
I've updated the [OP](https://github.com/dotnet/corefx/issues/19627#issue-227846040).
